osf_tests/management_commands/test_reindex_with_current_metrics_mappings.py

```diff
@@ -0,0 +1,112 @@
+
+from elasticsearch_metrics.field import Keyword
+
+from tests.json_api_test_app import JSONAPITestApp
+
+
+@pytest.fixture()
+def app():
+    return JSONAPITestApp()
+
+
+@pytest.mark.django_db
+class TestReindexingMetrics:
+
+    @pytest.fixture()
+    def preprint(self):
+        return PreprintFactory()
+
+    @pytest.fixture()
+    def user(self):
+        return AuthUserFactory()
+
+    @pytest.fixture()
+    def admin(self):
+        user = AuthUserFactory()
+        user.is_staff = True
+        user.add_system_tag('preprint_metrics')
+        user.save()
+        return user
+
+    @pytest.fixture()
+    def url(self):
+        return f'{settings.API_DOMAIN}_/metrics/preprints/downloads/'
+
+    @pytest.mark.es
+    def test_reindexing(self, app, url, preprint, user, admin, es6_client):
+        preprint_download = PreprintDownload.record_for_preprint(
+            preprint,
+            user,
+            version=1,
+            path='/MalcolmJenkinsKnockedBrandinCooksOutColdInTheSuperbowl',
+            random_new_field='Hi!'  # Here's our unmapped field! It's a text field by default.
+        )
+        preprint_download.save()
+
+        query = {
+            'aggs': {
+                'random_new_field': {
+                    'terms': {
+                        'field': 'random_new_field',  # Oh no, this is a text field, you can't query it like that!
+                    }
+                }
+            }
+        }
+
+        payload = {
+            'data': {
+                'type': 'preprint_metrics',
+                'attributes': {
+                    'query': query
+                }
+            }
+        }
+
+        # Hacky way to simulate a re-mapped index template
+        index_template = preprint_download._index
+        mapping = index_template._mapping
+        mapping.properties._params['properties']['random_new_field'] = Keyword(doc_values=True, index=True)
+        index_template._mapping._update_from_dict(mapping.to_dict())
+
+        # This should 400 because random_new_field is still stored as a text field despite the our index being remapped.
+        res = app.post_json_api(url, payload, auth=admin.auth, expect_errors=True)
+        assert res.status_code == 400
+        assert res.json['errors'][0]['detail'] == 'Fielddata is disabled on text fields by default. Set ' \
+                                                  'fielddata=true on [random_new_field] in order to load' \
+                                                  ' fielddata in memory by uninverting the inverted inde' \
+                                                  'x. Note that this can however use significant memory.' \
+                                                  ' Alternatively use a keyword field instead.'
+
+        call_command('reindex_with_current_metrics_mappings', f'--indices={preprint_download.meta["index"]}')
+        time.sleep(20)  # This is set for Travis, a 1-2 second latency should be good for testing locally.
```

Comment:
Is there no way to do this without a sleep? This is a) likely to fail probabilistically; and b) going to add 20 seconds to our test time.

Reply:
I'm disscussing this with @mfraezz now and it seems like with previous ES tests [We tried some stuff, but utimately gave up](https://github.com/CenterForOpenScience/osf.io/blob/develop/osf_tests/test_elastic_search.py#L699-L700)